### PR TITLE
better array handling

### PR DESCRIPTION
**Pull request by SimonCropp**
v3 and v4 still wrap single messages in an array

### CODE REVIEW

File: src/Tests/ArrayTests.cs [hunk 1]
```diff
@@ -41,8 +38,84 @@ public void Should_throw_for_array()
...
+    $type: '{typeName}',
+    SomeProperty: 'Value1'
+}}]";
+
+        using (var stream = new MemoryStream())
+        {
+            var streamWriter = new StreamWriter(stream);
+            streamWriter.Write(text);
+            streamWriter.Flush();
+            stream.Position = 0;
```
Comment (HEskandari) [inline src/Tests/ArrayTests.cs]: Put the sreamWriter stuff in a helper function for the test project? It just clutters the test code.

File: src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs [hunk 2]
```diff
@@ -69,43 +69,50 @@ public void Serialize(object message, Stream stream)
...
+                if (isArrayStream)
                 {
-                    jsonSerializer.Deserialize<object>(reader)
-                };
+                    var objects = (object[])jsonSerializer.Deserialize(reader, type.MakeArrayType());
+                    if (objects.Length > 1)
+                    {
+                        throw new Exception("Multiple messages in the same stream are not supported.");
+                    }
+                    return objects.Single();
```
Comment (HEskandari) [inline src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: Single or [0]? For me Single enforces one item being in the array, which is already checked for and is not necessary.

File: src/Tests/ArrayTests.cs (hunk 1 above)
Comment (SimonCropp) [inline src/Tests/ArrayTests.cs]: fixed

File: src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs (hunk 2 above)
Comment (SimonCropp) [inline src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: fixed

File: src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs [hunk 3]
```diff
@@ -69,43 +69,50 @@ public void Serialize(object message, Stream stream)
 
         public object[] Deserialize(Stream stream, IList<Type> messageTypes)
         {
-            if (IsArrayStream(stream))
-            {
-                throw new Exception("Multiple messages in the same stream are not supported.");
-            }
+            var isArrayStream = IsArrayStream(stream);
 
             if (messageTypes.Any())
```
Comment (danielmarbach) [inline src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: This could be replaced by Length which avoids LINQ

**Pull request merged**